### PR TITLE
Rational interval analysis

### PR DESCRIPTION
**Pull request by ajpal**
For some reason when I add tests they pass locally but then the build step fails in CI? Has anyone seen that before? I'll try to add tests as a separate PR to help isolate the issue

### CODE REVIEW

**Comment by bksaiki** [comment]
Chandra and I have been having trouble with this bug. For some odd reason the rust compiler will just sporadically fail. I suggest just re-running the CI job when this happens.

File: src/bin/rational.rs [hunk 1]
```diff
@@ -47,6 +49,41 @@ fn mk_constant(n: &BigInt, d: &BigInt) -> Option<Constant> {
...
+fn neg(interval: Interval<Constant>) -> Interval<Constant> {
+    (interval.1.map(|x| x.neg()), interval.0.map(|x| x.neg()))
+}
+
+fn abs(interval: Interval<Constant>) -> Interval<Constant> {
+    let (min, max) = interval;
+    let zero = Ratio::new(0.to_bigint().unwrap(), 1.to_bigint().unwrap());
+
+    let new_min = match (min.clone(), max.clone()) {
+        (None, None) => zero,
```
Comment (chandrakananandi) [inline src/bin/rational.rs:61]: what does `None` encode in these intervals? I thought it was `-inf` and `+inf` depending on which side of the interval it is on. Is that the case? We should add that in a comment if so.

Comment (chandrakananandi) [reply src/bin/rational.rs:61]: not here, but probably in `lib.rs` where Interval is defined.

**Review by chandrakananandi — CHANGES_REQUESTED** [review]
I left just a few suggestions for the data structure to use for `Interval` and questions about the use of `None` but looks good to me over all!

File: src/bin/rational.rs [hunk 2]
```diff
@@ -47,6 +49,41 @@ fn mk_constant(n: &BigInt, d: &BigInt) -> Option<Constant> {
...
+
+    let new_max = match (min, max) {
+        (Some(a), Some(b)) => Some(a.abs().max(b.abs())),
+        _ => None,
+    };
+    (Some(new_min), new_max)
+}
+
+fn add(a: Interval<Constant>, b: Interval<Constant>) -> Interval<Constant> {
+    let add = |(a, b): (Constant, Constant)| Some(a.add(b));
```
[the remark this answers is not in the document]
Comment (ajpal) [reply src/bin/rational.rs]: oh, I didn't realize `+` was overloaded :)

**Review by chandrakananandi — APPROVED** [review]

**Pull request merged**